pandas/tests/indexes/datetimes/test_date_range.py

```diff
@@ -1166,3 +1166,33 @@ def test_date_range_with_custom_holidays():
         freq=freq,
     )
     tm.assert_index_equal(result, expected)
+
+
+class TestDateRangeNonNano:
+    def test_date_range_reso_validation(self):
+        msg = "'reso' must be one of 's', 'ms', 'us', 'ns'"
+        with pytest.raises(ValueError, match=msg):
+            date_range("2016-01-01", "2016-03-04", periods=3, reso="h")
+
+    def test_date_range_freq_higher_than_reso(self):
+        # freq being higher-resolution than reso is a problem
+        msg = "Use a lower freq or a higher reso instead"
+        with pytest.raises(ValueError, match=msg):
+            # TODO give a more useful or informative message?
+            date_range("2016-01-01", "2016-01-01 00:00:00.000001", freq="ns", reso="ms")
+
+    def test_date_range_non_nano(self):
+        start = np.datetime64("1066-10-14")  # Battle of Hastings
+        end = np.datetime64("2305-07-13")  # Jean-Luc Picard's birthday
+
+        dti = date_range(start, end, freq="D", reso="s")
```

Comment:
Could you add a test where `freq` and `reso` have the same unit?

Also should `date_range` not raise if `start` and `end` have the same resolution but different from `reso`?

Also should `date_range` not raise if `start` and `end` have different resolutions?

Reply:
> Could you add a test where freq and reso have the same unit?

Sure

> Also should date_range not raise if start and end have the same resolution but different from reso?
> Also should date_range not raise if start and end have different resolutions?

Good question.  I don't think I'd want to raise if start and end have different resolutions bc if the user gives strings and we infer reso for those (which we dont do yet) they can easily get different resos.

How about we do `ts._as_unit(unit, round_ok=False)` so it Just Works as long as the casting can be non-lossy, raises otherwise?

Reply:
How about we do ts._as_unit(unit, round_ok=False) so it Just Works as long as the casting can be non-lossy, raises otherwise?

Sure that sounds reasonable